src/parser/mod.rs

```diff
@@ -0,0 +1,695 @@
+				Payload::StartSection { func, range } => {
+					info.start_function = Some(func);
+					info.section(SectionId::Start.into(), range, input_wasm);
+				},
+				Payload::ElementSection(reader) => {
+					info.elements_count = reader.count();
+					info.section(SectionId::Element.into(), reader.range(), input_wasm);
+				},
+				Payload::DataSection(reader) => {
+					info.data_segments_count = reader.count();
+					info.section(SectionId::Data.into(), reader.range(), input_wasm);
+				},
+				#[allow(unused_variables)]
+				Payload::CustomSection(c) => {
+					// FIXME
+					//  We are skipping the section due to following reason:
+					//  when dumping with custom section, WASM file is invalid:
+					//   'invalid UTF-8 encoding (at offset 0xd)'
+					//  Most likely it needs different encoding
+					//  To reproduce, call validate() with "custom_section_parse" feature enabled
+
+					#[cfg(feature = "custom_section_parse")]
+					// At the moment only name section supported
+					if c.name() == "name" {
+						info.section(
+							SectionId::Custom.into(),
+							Range { start: c.data_offset(), end: c.range().end },
+							input_wasm,
+						);
+					}
+				},
+
+				Payload::DataCountSection { count: _, range } => {
+					info.section(SectionId::DataCount.into(), range, input_wasm);
+				},
+				Payload::Version { .. } => {},
+				Payload::End(_) => {
+					break;
+				},
+				_ => todo!("{:?} not implemented", payload),
+			}
+			wasm = &wasm[consumed..];
+		}
+
+		Ok(info)
+	}
+
+	/// Validates the WASM binary
+	pub fn validate(&self, features: WasmFeatures) -> Result<()> {
+		// TODO validate_all() creates internal parser and parses the binary again. Rework to validate while parsing
+		Validator::new_with_features(features)
+			.validate_all(&self.bytes())
+			.map(|_| ())
+			.map_err(|err| anyhow!(err))
+	}
+
+	/// Registers a new raw_section in the ModuleInfo
+	pub fn section(&mut self, id: u8, range: Range<usize>, full_wasm: &[u8]) {
+		self.raw_sections.insert(id, RawSection::new(id, full_wasm[range].to_vec()));
+	}
+
+	/// Returns the function type based on the index of the type
+	/// `types[idx]`
+	pub fn get_type_by_idx(&self, type_idx: u32) -> Result<&Type> {
+		if type_idx >= self.types_map.len() as u32 {
+			return Err(anyhow!("type {} does not exist", type_idx));
+		}
+		Ok(&self.types_map[type_idx as usize])
+	}
+
+	/// Returns the function type based on the index of the function
+	/// `types[functions[idx]]`
+	pub fn get_type_by_func_idx(&self, func_idx: u32) -> Result<&Type> {
+		if func_idx >= self.function_map.len() as u32 {
+			return Err(anyhow!("function {} does not exist", func_idx));
+		}
+		let type_idx = self.function_map[func_idx as usize];
+		self.get_type_by_idx(type_idx)
+	}
+
+	pub fn resolve_type_idx(&self, t: &Type) -> Option<u32> {
+		let Type::Func(dt) = t else { todo!() };
+		for (index, ty) in self.types_map.iter().enumerate() {
+			let Type::Func(ot) = ty else { todo!() };
+			if ot.eq(dt) {
```

Comment:
Why not support both `Type::Func` and `Type::Array` by comparing `Type`.

Reply:
Type::Array is a part of [GC proposal](https://github.com/WebAssembly/gc), which is being in [implementation phase](https://github.com/WebAssembly/proposals/blob/main/README.md). 
But sure, I can add support.
But cannot compare `Type` directly, since it does not derive `PartialEq`/`Eq`.

Reply:
Please note, that there are also other places, where Array type support could be added. 
eg. 
- https://github.com/radixdlt/wasm-instrument/blob/refactor/drop_parity-wasm_dependency/src/util/translator.rs#L157
- https://github.com/radixdlt/wasm-instrument/blob/refactor/drop_parity-wasm_dependency/src/stack_limiter/thunk.rs#L71
- https://github.com/radixdlt/wasm-instrument/blob/refactor/drop_parity-wasm_dependency/src/stack_limiter/max_height.rs#L141

So maybe it would make sense to do it in separate PR?